File: 6- Karatsuba algorithm for fast multiplication/Karatsuba.java
public class Karatsuba {
    public static long multiply(long x, long y) {
        if (x < 10 && y < 10) {
            return x * y;
        }
        int noOneLength = numLength(x);
        int noTwoLength = numLength(y);
        int maxNumLength
                = Math.max(noOneLength, noTwoLength);
        Integer halfMaxNumLength
                = (maxNumLength / 2) + (maxNumLength % 2);
        long maxNumLengthTen
                = (long) Math.pow(10, halfMaxNumLength);
        long a = x / maxNumLengthTen;
        long b = x % maxNumLengthTen;
        long c = y / maxNumLengthTen;
        long d = y % maxNumLengthTen;
        long z0 = multiply(a, c);
        long z1 = multiply(a + b, c + d);
        long z2 = multiply(b, d);
        long ans = (z0 * (long) Math.pow(10, halfMaxNumLength * 2) + ((z1 - z0 - z2) * (long) Math.pow(10, halfMaxNumLength) + z2));
        return ans;
    }
    public static long  multiply(long x,long y,int BASE){
        int xLength = Long.toString(x).length();
        int yLength = Long.toString(y).length();
        int max = Math.max(xLength,yLength);
        if(max < 10)
            return (long)x*(long)y;
        int N = max/2;
        long multiplier = (long)Math.pow(BASE,N);
        long a = x/multiplier;
        long b = x%multiplier;
        long c = y/multiplier;
        long d = y%multiplier;
        long ac = multiply(a,c,BASE);
        long bd = multiply(b,d,BASE);
        long res = multiply((a+b),(c+d),BASE);
        long result = ac * (long)Math.pow(BASE,N)+ bd+((long)Math.pow(BASE,N/2)*res);
        return result;
    }

    private static int numLength(long n) {
        int noLen = 0;
        while (n > 0) {
            noLen++;
            n /= 10;
        }
        return noLen;
    }
}

File: Mathematical Algorithms/Karatsuba algorithm for fast multiplication/Karatsuba.cpp
#include <vector>
#include <cstdlib>
#include <cmath>
#include <algorithm>
using namespace std;
class Karatsuba {
public:
    static long long Multiply(long long U, long long V, int n){
        if (U==0 || V==0)
            return 0;
        else if (n<=5)
            return U*V;
        if (n%2!=0)
            n=n+1;

        long long x=(long long)(U/pow(10,n/2));
        long long y=(U%(long long)(pow(10,n/2)));
        long long w=(long long)(V/pow(10,n/2));
        long long z=(V%(long long)(pow(10,n/2)));

        long long xw=Multiply(x,w,n/2);
        long long yz=Multiply(y,z,n/2);
        long long xywz = Multiply((x+y),(w+z),n/2)-xw-yz;

        return (long long)(xw*pow(10,n)+xywz*pow(10,n/2)+yz);
    }
};

